### ISSUE TITLE
Increasing batch size leads to decrease in speed (increase in time) 

### ISSUE THREAD

**Issue by sajastu** [OpenNMT/OpenNMT-py#1391]
Hi,

I have been training Transformer model with different batch sizes. Each time, I got different time intervals to train each 50 training steps. With larger batch sizes, I expected the interval to decrease.

Here is the results for -batch_size 4096 and -max_generator_batches 32:

> [2019-04-09 19:19:06,807 INFO] Step 50/200000; acc:   1.84; ppl: 3041.89; xent: 8.02; lr: 0.00000; 15693/3803 tok/s;     12 sec
> [2019-04-09 19:19:13,992 INFO] Step 100/200000; acc:   2.48; ppl: 2701.39; xent: 7.90; lr: 0.00001; 26261/7826 tok/s;     20 sec
> [2019-04-09 19:19:21,326 INFO] Step 150/200000; acc:   2.77; ppl: 2948.40; xent: 7.99; lr: 0.00001; 25761/7146 tok/s;     27 sec
> [2019-04-09 19:19:28,536 INFO] Step 200/200000; acc:   3.71; ppl: 2655.22; xent: 7.88; lr: 0.00001; 26035/7397 tok/s;     34 sec
> ...

As shown, it seems that every 50 training steps takes about 7 seconds to be trained.

However, when I put 64 as the batch_size and -max_generator_batches 32, it takes about 3 seconds to be trained.

> [2019-04-09 19:21:26,847 INFO] Step 50/200000; acc:   2.32; ppl: 2037.73; xent: 7.62; lr: 0.00000; 614/168 tok/s;      8 sec
> [2019-04-09 19:21:29,699 INFO] Step 100/200000; acc:   2.40; ppl: 3013.75; xent: 8.01; lr: 0.00001; 2036/598 tok/s;     11 sec
> [2019-04-09 19:21:32,607 INFO] Step 150/200000; acc:   2.23; ppl: 2500.22; xent: 7.82; lr: 0.00001; 2378/677 tok/s;     14 sec
> [2019-04-09 19:21:35,473 INFO] Step 200/200000; acc:   3.19; ppl: 2486.73; xent: 7.82; lr: 0.00001; 1822/590 tok/s;     17 sec
> [2019-04-09 19:21:38,402 INFO] Step 250/200000; acc:   2.29; ppl: 2871.41; xent: 7.96; lr: 0.00002; 2232/552 tok/s;     20 sec
> ...
> 

I also experimented with different values of `-max_generator_batches` but that did not help. Actually, the time interval was decreasing until some fixed `-max_generator_batches`, but then not changed even with larger values of `-max_generator_batches`.

Could you please explain this if I'm not wrong? Because it does not make sense to me.

This is the command I'm running for the training:

`python train.py -save_model models/star_layer4_lr02 -data ~/packages/openNMT/data/med/MED -word_vec_size 100  -encoder_type transformer -decoder_type transformer -pre_word_vecs_enc "data/embeddings.enc.pt" -normalization tokens -pre_word_vecs_dec "data/embeddings.dec.pt" -train_steps 200000 -dropout 0.1  -optim adam -learning_rate 0.2 -batch_type tokens -copy_attn -param_init 0 -param_init_glorot  -seed 1234 -transformer_ff 2048 -heads 5 -position_encoding -share_embeddings -adam_beta1 0.9 -adam_beta2 0.998 -label_smoothing 0.1 -decay_method noam -warmup_steps 4000 -rnn_size 100 -layers 4 -batch_size 64 -max_generator_batches 32 -world_size 1 -gpu_ranks 0`

Thanks!

**Comment by vince62s** [comment]
try to better understand things.
batch_type token => 64 tokens is really low, and yes it takes less time to process 64 tokens vs 4096 tokens.
but you will see the speed tokens / sec which is way higher with 4096.
